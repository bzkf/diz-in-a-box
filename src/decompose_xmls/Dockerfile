# syntax=docker/dockerfile:1

# Comments are provided throughout this file to help you get started.
FROM docker.io/library/python:3.11.3-slim@sha256:eaee5f73efa9ae962d2077756292bc4878c04fcbc13dc168bb00cc365f35647e AS base
ENV PYTHONDONTWRITEBYTECODE=1 \
    PYTHONUNBUFFERED=1
WORKDIR /app

# Create a non-privileged user that the app will run under.
# See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user
ARG UID=10001
RUN adduser \
    --disabled-password \
    --gecos "" \
    --home "/nonexistent" \
    --shell "/sbin/nologin" \
    --no-create-home \
    --uid "${UID}" \
    appuser

# Download dependencies as a separate step to take advantage of Docker's caching.
# Leverage a cache mount to /root/.cache/pip to speed up subsequent builds.
# Leverage a bind mount to requirements.txt to avoid having to copy them into
# into this layer.
RUN --mount=type=cache,target=/root/.cache/pip \
    --mount=type=bind,source=requirements.txt,target=requirements.txt \
    python -m pip install --no-cache-dir -r requirements.txt

RUN chown -R appuser:appuser /app

# Copy the source code into the container.
COPY decompose_xmls.py /app

FROM base AS test
RUN --mount=type=cache,target=/root/.cache/pip \
    --mount=type=bind,source=requirements-test.txt,target=requirements-test.txt \
    python -m pip install --no-cache-dir -r requirements-test.txt

COPY . .

RUN PYTHONPATH=${PWD} pytest -vv --cov

FROM base
# Switch to the non-privileged user to run the application.
USER 10001

# Run the application.
CMD ["python", "/app/decompose_xmls.py"]
