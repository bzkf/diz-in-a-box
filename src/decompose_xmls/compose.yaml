services:
  decompose-xmls:
    build:
      context: .
    user: "${UID:-1000}:${GID:-1000}"
    volumes:
      - ./input-adt-reports:/app/input-adt-reports
      - ./output-adt-reports:/app/output-files
