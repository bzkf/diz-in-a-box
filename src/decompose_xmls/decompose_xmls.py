import json
import os
import time
import xml.etree.ElementTree as ET
from dataclasses import dataclass

from confluent_kafka import Producer
from pydantic import BaseSettings


class Settings(BaseSettings):
    input_folder: str = "./input-adt-reports"
    output_folder: str = "./output-files"
    output_folder_xml: str = "./output-files/output-xmls"
    save_as_files_enabled: bool = True
    kafka_enabled: bool = False
    kafka_bootstrap_servers: str = "localhost:9092"
    kafka_output_topic: str = "adt.einzelmeldungen"


@dataclass
class Einzelmeldung:
    xml: ET.Element
    patient_id: str
    meldung_id: str

    def __repr__(self) -> str:
        ET.register_namespace("", "http://www.gekid.de/namespace")
        dict_repr = {
            "xml": ET.tostring(self.xml, encoding="unicode"),
            "patient_id": self.patient_id,
            "meldung_id": self.meldung_id,
        }
        return json.dumps(dict_repr)


def conditional_folder_create(folder_name):
    if not os.path.exists(folder_name):
        os.makedirs(folder_name)


def save_xml_files(meldung_root, patient_id, meldung_id):
    settings = Settings()
    ET.register_namespace("", "http://www.gekid.de/namespace")
    ET.ElementTree(meldung_root).write(
        f"{settings.output_folder_xml}/patient_{patient_id}_meldung_{meldung_id}.xml",
        encoding="UTF-8",
        xml_declaration=True,
    )


def decompose_sammelmeldung(root: ET.Element, filename: str) -> list[Einzelmeldung]:
    # Get all "Patient" elements, save the absender to be appended to each new file
    patients = root.findall(".//{http://www.gekid.de/namespace}Patient")
    absender = root.find("./{http://www.gekid.de/namespace}Absender")

    if absender is None:
        print(f"Absender is not defined in the input file {filename}. Stopping.")
        return []

    result: list[Einzelmeldung] = []

    # Loop through each patient
    for patient in [p for p in patients if p is not None]:
        # Get the patient ID - might remove from filename later
        # and only keep the meldung_id
        if (
            patient_id_element := patient.find(
                ".//{http://www.gekid.de/namespace}Patienten_Stammdaten"
            )
        ) is None:
            print(f"Patienten_Stammdaten is unset for {filename}. Skipping.")
            continue

        patient_id = patient_id_element.get("Patient_ID")
        if patient_id is None:
            print(f"Patient_ID is unset for {filename}. Skipping.")
            continue

        # Get all "Meldung" elements for this patient
        meldungen = patient.findall(".//{http://www.gekid.de/namespace}Meldung")

        # Loop through each meldung for this patient
        for meldung in meldungen:
            meldung_id = meldung.get("Meldung_ID")
            if meldung_id is None:
                print(f"Meldung_ID is unset for {filename}. Skipping.")
                continue

            # NEW IDEA - copy root, LOOP THROUGH EXISTING MELDUNGEN, BUT ONLY KEEP
            # THE ONE WITH meldung_id and remove the rest
            meldung_root = ET.Element(root.tag, root.attrib)
            meldung_root.append(absender)

            # reintroduce parent tag "Menge_Patient" that gets lost in the looping
            menge_patient = ET.Element("Menge_Patient")
            menge_patient.append(patient)
            meldung_root.append(menge_patient)

            menge_meldung_element = meldung_root.find(
                ".//{http://www.gekid.de/namespace}Menge_Meldung"
            )

            if menge_meldung_element is None:
                print(f"Menge_Meldung element not found in {filename}")
                continue

            menge_meldung_element.append(meldung)

            relevant_meldung = None
            # loop through the newly build meldung_root and remove all meldungen
            # with meldung_id not matching to currrent one and remove duplicates
            for einzelmeldung in meldung_root.findall(
                ".//{http://www.gekid.de/namespace}Meldung"
            ):
                if einzelmeldung.attrib["Meldung_ID"] == meldung_id:
                    if relevant_meldung is None:
                        relevant_meldung = einzelmeldung
                    else:
                        menge_meldung_element.remove(einzelmeldung)
                else:
                    menge_meldung_element.remove(einzelmeldung)

            result.append(Einzelmeldung(meldung_root, patient_id, meldung_id))

    return result


def kafka_delivery_report(err, msg):
    if err is not None:
        print(f"Message delivery failed: {err}")
    else:
        print(f"Message delivered to {msg.topic()} [{msg.partition()}]")


def decompose_folder(input_folder: str):
    settings = Settings()

    kafka_producer: Producer = None
    if settings.kafka_enabled:
        kafka_producer = Producer(
            {"bootstrap.servers": settings.kafka_bootstrap_servers}
        )

    for xmlfile in os.listdir(input_folder):
        if not xmlfile.endswith(".xml"):
            continue
        filename = os.path.join(input_folder, xmlfile)
        tree = ET.parse(filename)
        root = tree.getroot()

        for einzelmeldung in decompose_sammelmeldung(root, filename):
            # saves json files for kafka bridge input in this schema
            xml_str = ET.tostring(root, encoding="unicode")

            # prepare json files for kafka bridge
            result_data = {
                "LKR_MELDUNG": einzelmeldung.meldung_id,
                "XML_DATEN": xml_str,
                "VERSIONSNUMMER": 1,
                "REFERENZ_NUMMER": einzelmeldung.patient_id,
            }

            if settings.save_as_files_enabled:
                with open(
                    f"{settings.output_folder}/"
                    + f"patient_{einzelmeldung.patient_id}_meldung_"
                    + f"{einzelmeldung.meldung_id}.json",
                    "w",
                    encoding="utf-8",
                ) as f:
                    json.dump(result_data, f, indent=4)

                save_xml_files(
                    einzelmeldung.xml,
                    einzelmeldung.patient_id,
                    einzelmeldung.meldung_id,
                )

            if kafka_producer is not None:
                kafka_producer.poll(0)

                # Asynchronously produce a message. The delivery report callback will
                # be triggered from the call to poll() above, or flush() below, when the
                # message has been successfully delivered or failed permanently.
                kafka_producer.produce(
                    settings.kafka_output_topic,
                    json.dumps(result_data),
                    callback=kafka_delivery_report,
                    key=f"{einzelmeldung.patient_id}-{einzelmeldung.meldung_id}",
                )

    if kafka_producer is not None:
        kafka_producer.flush()


def main():
    start = time.monotonic()
    settings = Settings()
    # do all stuff here
    conditional_folder_create(settings.output_folder_xml)
    decompose_folder(settings.input_folder)

    end = time.monotonic()
    print(f"time elapsed: {end - start}s")


if __name__ == "__main__":
    main()
