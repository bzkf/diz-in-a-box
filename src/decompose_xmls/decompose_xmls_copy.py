import json
import os
import time
import xml.etree.ElementTree as ET
from dataclasses import dataclass
from io import BytesIO

from confluent_kafka import Producer
from pydantic import BaseSettings


class Settings(BaseSettings):
    input_folder: str = "./input-obds-reports"
    output_folder: str = "./output-files"   # RENAME THIS
    output_folder_xml: str = "./output-files/output-xmls"
    save_as_files_enabled: bool = True
    kafka_enabled: bool = False
    kafka_bootstrap_servers: str = "localhost:9092"
    kafka_output_topic: str = "obds.einzelmeldungen"


def conditional_folder_create(folder_name):
    if not os.path.exists(folder_name):
        os.makedirs(folder_name)


def save_xml_files(meldung_root, patient_id, meldung_id):
    settings = Settings()
    ET.register_namespace("", "http://www.gekid.de/namespace")
    tree = ET.ElementTree(meldung_root)
    ET.indent(tree, "  ")
    # ET.ElementTree(meldung_root)
    tree.write(
        f"{settings.output_folder_xml}/patient_{patient_id}"
        f"_meldung_{meldung_id}.xml",
        encoding="UTF-8",
        xml_declaration=True,
    )


def kafka_delivery_report(err, msg):
    if err is not None:
        print(f"Message delivery failed: {err}")
    else:
        print(f"Message delivered to {msg.topic()}@{msg.partition()}")


def decompose_sammelmeldung(root: ET.Element, filename: str):
    settings = Settings()
    kafka_producer: Producer = None
    if settings.kafka_enabled:
        kafka_producer = Producer(
            {"bootstrap.servers": settings.kafka_bootstrap_servers}
        )
    # Get all "Patient" elements, save the absender to be appended to each new file - IN THE END
    patients = root.findall(".//{http://www.gekid.de/namespace}Patient")
    absender = root.find("./{http://www.gekid.de/namespace}Absender")

    if absender is None:
        print(f"Absender is not defined in the input file {filename}. Stopping.")
        return []

    menge_patient = ET.Element("Menge_Patient")

    # Loop through each patient
    for patient in [p for p in patients if p is not None]:
        if (
            patient_id_element := patient.find(
                ".//{http://www.gekid.de/namespace}Patienten_Stammdaten"
            )
        ) is None:
            print(f"Patienten_Stammdaten is unset for {filename}. Skipping.")
            continue

        patient_id = patient_id_element.get("Patient_ID")
        if patient_id is None:
            print(f"Patient_ID is unset for {filename}. Skipping.")
            continue

        # remove all Menge_Meldung
        menge_meldung = patient.find('.//{http://www.gekid.de/namespace}Menge_Meldung')
        if menge_meldung is not None:
            patient.remove(menge_meldung)

        for meldung in menge_meldung:
            # get meldung_id
            meldung_id = meldung.get("Meldung_ID")
            menge_meldung_group = ET.Element("Menge_Meldung")
            # APPEND IN THE END
            menge_meldung_group.append(meldung)
            #menge_patient = ET.Element("Menge_Patient")
            menge_patient.append(patient)
            menge_patient.append(menge_meldung_group)
            meldung_root = ET.Element(root.tag, root.attrib)
            meldung_root.append(absender)
            meldung_root.append(menge_patient)

            ET.register_namespace("", "http://www.gekid.de/namespace")
            f = BytesIO()
            tree = ET.ElementTree(meldung_root)
            ET.indent(tree, "  ")
            tree.write(
                f, encoding="utf-8", xml_declaration=True
            )

            xml_str = f.getvalue().decode()
            # prepare json files for kafka bridge
            result_data = {
                "payload": {
                    "LKR_MELDUNG": meldung_id,
                    "XML_DATEN": xml_str,
                    "VERSIONSNUMMER": 1,
                    "REFERENZ_NUMMER": patient_id,
                }
            }

            if settings.save_as_files_enabled:
                with open(
                    f"{settings.output_folder}/"
                    + f"patient_{patient_id}_meldung_"
                    + f"{meldung_id}.json",
                    "w",
                    encoding="utf-8",
                ) as output_file:
                    json.dump(result_data, output_file, indent=4)

            save_xml_files(
                    meldung_root,
                    patient_id,
                    meldung_id,
                )

            if kafka_producer is not None:
                kafka_producer.poll(0)

                # Asynchronously produce a message. The delivery report callback will
                # be triggered from the call to poll() above, or flush() below, when the
                # message has been successfully delivered or failed permanently.
                kafka_producer.produce(
                    settings.kafka_output_topic,
                    json.dumps(result_data),
                    callback=kafka_delivery_report,
                    key=f"{patient_id}-{meldung_id}",
                )

    if kafka_producer is not None:
        kafka_producer.flush()


def decompose_folder(input_folder: str):
    settings = Settings()

    if settings.save_as_files_enabled:
        conditional_folder_create(settings.output_folder_xml)

    for xmlfile in os.listdir(input_folder):
        if not xmlfile.endswith(".xml"):
            continue
        filename = os.path.join(input_folder, xmlfile)
        tree = ET.parse(filename)
        root = tree.getroot()

        decompose_sammelmeldung(root, filename)


def main():
    start = time.monotonic()
    settings = Settings()

    decompose_folder(settings.input_folder)

    end = time.monotonic()
    print(f"time elapsed: {end - start}s")


if __name__ == "__main__":
    main()
