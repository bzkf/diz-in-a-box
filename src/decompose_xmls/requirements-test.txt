pytest==7.4.3
syrupy==4.6.0
pytest-cov==4.1.0
