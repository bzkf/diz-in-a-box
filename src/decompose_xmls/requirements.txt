pydantic==1.10.14
confluent-kafka==2.3.0
