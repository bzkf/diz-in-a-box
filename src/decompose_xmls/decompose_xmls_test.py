import xml.etree.ElementTree as ET

import pytest

from decompose_xmls.decompose_xmls import decompose_sammelmeldung


@pytest.mark.parametrize(
    "adt_input_file_path", [("input-adt-reports/test-2patients.xml")]
)
def test_decompose_sammelmeldung(snapshot, adt_input_file_path):
    tree = ET.parse(adt_input_file_path)
    root = tree.getroot()

    result = decompose_sammelmeldung(root, adt_input_file_path)
    assert result == snapshot
