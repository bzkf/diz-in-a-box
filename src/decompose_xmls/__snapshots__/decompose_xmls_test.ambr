# serializer version: 1
# name: test_decompose_sammelmeldung[input-adt-reports/test-2patients.xml]
  list([
    {"xml": "<?xml version='1.0' encoding='utf-8'?>\n<ADT_GEKID xmlns=\"http://www.gekid.de/namespace\" Schema_Version=\"2.2.1\"><Absender Absender_ID=\"UKKK_ONKOSTAR\" Software_ID=\"ONKOSTAR\" Installations_ID=\"2.11.1.4\">\n    <Absender_Bezeichnung>UKKK</Absender_Bezeichnung>\n    <Absender_Ansprechpartner />\n    <Absender_Anschrift>xxxxxxxxxxxxxxxxxxxx</Absender_Anschrift>\n  </Absender>\n  <Menge_Patient><Patient>\n      <Patienten_Stammdaten Patient_ID=\"1004444444\">\n        <KrankenversichertenNr>F00000000</KrankenversichertenNr>\n        <KrankenkassenNr>9999999</KrankenkassenNr>\n        <Patienten_Nachname>Hafnero</Patienten_Nachname>\n        <Patienten_Vornamen>B</Patienten_Vornamen>\n        <Patienten_Geburtsname>E</Patienten_Geburtsname>\n        <Patienten_Geschlecht>W</Patienten_Geschlecht>\n        <Patienten_Geburtsdatum>14.01.1900</Patienten_Geburtsdatum>\n        <Menge_Adresse>\n          <Adresse>\n            <Patienten_Land>DE</Patienten_Land>\n            <Patienten_PLZ>91058</Patienten_PLZ>\n            <Patienten_Ort>Erlangen</Patienten_Ort>\n          </Adresse>\n        </Menge_Adresse>\n      </Patienten_Stammdaten>\n      <Menge_Meldung>\n        <Meldung Meldung_ID=\"FK0037276\" Melder_ID=\"FK\">\n          <Meldedatum>10.03.1900</Meldedatum>\n          <Meldebegruendung>I</Meldebegruendung>\n          <Meldeanlass>behandlungsende</Meldeanlass>\n          <Tumorzuordnung Tumor_ID=\"1\">\n            <Primaertumor_ICD_Code>D07.3</Primaertumor_ICD_Code>\n            <Primaertumor_ICD_Version>10 2022 GM</Primaertumor_ICD_Version>\n            <Diagnosedatum>08.08.1900</Diagnosedatum>\n            <Seitenlokalisation>L</Seitenlokalisation>\n          </Tumorzuordnung>\n          <Menge_OP>\n            <OP OP_ID=\"280098\">\n              <OP_Intention>K</OP_Intention>\n              <OP_Datum>01.09.2022</OP_Datum>\n              <Menge_OPS>\n                <OP_OPS>5-543.42</OP_OPS>\n                <OP_OPS>5-683.00</OP_OPS>\n              </Menge_OPS>\n              <OP_OPS_Version>2022</OP_OPS_Version>\n              <Histologie Histologie_ID=\"280094\">\n                <Tumor_Histologiedatum>01.09.1900</Tumor_Histologiedatum>\n                <Histologie_EinsendeNr>12226123</Histologie_EinsendeNr>\n                <Morphologie_Code>8460/2</Morphologie_Code>\n                <Morphologie_ICD_O_Version>33</Morphologie_ICD_O_Version>\n                <Grading>B</Grading>\n              </Histologie>\n              <TNM TNM_ID=\"280117_d\">\n                <TNM_Datum>01.09.2022</TNM_Datum>\n                <TNM_Version>8</TNM_Version>\n                <TNM_c_p_u_Praefix_T>p</TNM_c_p_u_Praefix_T>\n                <TNM_T>1c3</TNM_T>\n                <TNM_c_p_u_Praefix_N>c</TNM_c_p_u_Praefix_N>\n                <TNM_N>0</TNM_N>\n                <TNM_c_p_u_Praefix_M>c</TNM_c_p_u_Praefix_M>\n                <TNM_M>0</TNM_M>\n                <TNM_L>L0</TNM_L>\n                <TNM_V>V0</TNM_V>\n                <TNM_Pn>Pn0</TNM_Pn>\n              </TNM>\n              <Residualstatus>\n                <Lokale_Beurteilung_Residualstatus>R0</Lokale_Beurteilung_Residualstatus>\n                <Gesamtbeurteilung_Residualstatus>R0</Gesamtbeurteilung_Residualstatus>\n              </Residualstatus>\n              <Menge_Komplikation>\n                <OP_Komplikation>N</OP_Komplikation>\n              </Menge_Komplikation>\n              <Menge_Operateur>\n                <Name_Operateur>jjjjj</Name_Operateur>\n                <Name_Operateur>xxxxx</Name_Operateur>\n              </Menge_Operateur>\n              <Anmerkung>Makroskopischer Tumorrest: R0</Anmerkung>\n            </OP>\n          </Menge_OP>\n        </Meldung>\n      </Menge_Meldung>\n    </Patient>\n    </Menge_Patient></ADT_GEKID>", "patient_id": "1004444444", "meldung_id": "FK0037275"},
    {"xml": "<?xml version='1.0' encoding='utf-8'?>\n<ADT_GEKID xmlns=\"http://www.gekid.de/namespace\" Schema_Version=\"2.2.1\"><Absender Absender_ID=\"UKKK_ONKOSTAR\" Software_ID=\"ONKOSTAR\" Installations_ID=\"2.11.1.4\">\n    <Absender_Bezeichnung>UKKK</Absender_Bezeichnung>\n    <Absender_Ansprechpartner />\n    <Absender_Anschrift>xxxxxxxxxxxxxxxxxxxx</Absender_Anschrift>\n  </Absender>\n  <Menge_Patient><Patient>\n      <Patienten_Stammdaten Patient_ID=\"1004444444\">\n        <KrankenversichertenNr>F00000000</KrankenversichertenNr>\n        <KrankenkassenNr>9999999</KrankenkassenNr>\n        <Patienten_Nachname>Hafnero</Patienten_Nachname>\n        <Patienten_Vornamen>B</Patienten_Vornamen>\n        <Patienten_Geburtsname>E</Patienten_Geburtsname>\n        <Patienten_Geschlecht>W</Patienten_Geschlecht>\n        <Patienten_Geburtsdatum>14.01.1900</Patienten_Geburtsdatum>\n        <Menge_Adresse>\n          <Adresse>\n            <Patienten_Land>DE</Patienten_Land>\n            <Patienten_PLZ>91058</Patienten_PLZ>\n            <Patienten_Ort>Erlangen</Patienten_Ort>\n          </Adresse>\n        </Menge_Adresse>\n      </Patienten_Stammdaten>\n      <Menge_Meldung>\n        <Meldung Meldung_ID=\"FK0037276\" Melder_ID=\"FK\">\n          <Meldedatum>10.03.1900</Meldedatum>\n          <Meldebegruendung>I</Meldebegruendung>\n          <Meldeanlass>behandlungsende</Meldeanlass>\n          <Tumorzuordnung Tumor_ID=\"1\">\n            <Primaertumor_ICD_Code>D07.3</Primaertumor_ICD_Code>\n            <Primaertumor_ICD_Version>10 2022 GM</Primaertumor_ICD_Version>\n            <Diagnosedatum>08.08.1900</Diagnosedatum>\n            <Seitenlokalisation>L</Seitenlokalisation>\n          </Tumorzuordnung>\n          <Menge_OP>\n            <OP OP_ID=\"280098\">\n              <OP_Intention>K</OP_Intention>\n              <OP_Datum>01.09.2022</OP_Datum>\n              <Menge_OPS>\n                <OP_OPS>5-543.42</OP_OPS>\n                <OP_OPS>5-683.00</OP_OPS>\n              </Menge_OPS>\n              <OP_OPS_Version>2022</OP_OPS_Version>\n              <Histologie Histologie_ID=\"280094\">\n                <Tumor_Histologiedatum>01.09.1900</Tumor_Histologiedatum>\n                <Histologie_EinsendeNr>12226123</Histologie_EinsendeNr>\n                <Morphologie_Code>8460/2</Morphologie_Code>\n                <Morphologie_ICD_O_Version>33</Morphologie_ICD_O_Version>\n                <Grading>B</Grading>\n              </Histologie>\n              <TNM TNM_ID=\"280117_d\">\n                <TNM_Datum>01.09.2022</TNM_Datum>\n                <TNM_Version>8</TNM_Version>\n                <TNM_c_p_u_Praefix_T>p</TNM_c_p_u_Praefix_T>\n                <TNM_T>1c3</TNM_T>\n                <TNM_c_p_u_Praefix_N>c</TNM_c_p_u_Praefix_N>\n                <TNM_N>0</TNM_N>\n                <TNM_c_p_u_Praefix_M>c</TNM_c_p_u_Praefix_M>\n                <TNM_M>0</TNM_M>\n                <TNM_L>L0</TNM_L>\n                <TNM_V>V0</TNM_V>\n                <TNM_Pn>Pn0</TNM_Pn>\n              </TNM>\n              <Residualstatus>\n                <Lokale_Beurteilung_Residualstatus>R0</Lokale_Beurteilung_Residualstatus>\n                <Gesamtbeurteilung_Residualstatus>R0</Gesamtbeurteilung_Residualstatus>\n              </Residualstatus>\n              <Menge_Komplikation>\n                <OP_Komplikation>N</OP_Komplikation>\n              </Menge_Komplikation>\n              <Menge_Operateur>\n                <Name_Operateur>jjjjj</Name_Operateur>\n                <Name_Operateur>xxxxx</Name_Operateur>\n              </Menge_Operateur>\n              <Anmerkung>Makroskopischer Tumorrest: R0</Anmerkung>\n            </OP>\n          </Menge_OP>\n        </Meldung>\n      </Menge_Meldung>\n    </Patient>\n    </Menge_Patient></ADT_GEKID>", "patient_id": "1004444444", "meldung_id": "FK0037276"},
    {"xml": "<?xml version='1.0' encoding='utf-8'?>\n<ADT_GEKID xmlns=\"http://www.gekid.de/namespace\" Schema_Version=\"2.2.1\"><Absender Absender_ID=\"UKKK_ONKOSTAR\" Software_ID=\"ONKOSTAR\" Installations_ID=\"2.11.1.4\">\n    <Absender_Bezeichnung>UKKK</Absender_Bezeichnung>\n    <Absender_Ansprechpartner />\n    <Absender_Anschrift>xxxxxxxxxxxxxxxxxxxx</Absender_Anschrift>\n  </Absender>\n  <Menge_Patient><Patient>\n      <Patienten_Stammdaten Patient_ID=\"101777777777\">\n        <KrankenversichertenNr>277777777</KrankenversichertenNr>\n        <KrankenkassenNr>9777777777</KrankenkassenNr>\n        <Patienten_Nachname>Riedelo</Patienten_Nachname>\n        <Patienten_Titel />\n        <Patienten_Namenszusatz />\n        <Patienten_Vornamen>M</Patienten_Vornamen>\n        <Patienten_Geburtsname />\n        <Patienten_Geschlecht>M</Patienten_Geschlecht>\n        <Patienten_Geburtsdatum>28.10.1900</Patienten_Geburtsdatum>\n        <Menge_Adresse>\n          <Adresse>\n            <Patienten_Strasse>xxxxxxxxx</Patienten_Strasse>\n            <Patienten_Hausnummer />\n            <Patienten_Land>DE</Patienten_Land>\n            <Patienten_PLZ>90</Patienten_PLZ>\n            <Patienten_Ort>xxxxxxx</Patienten_Ort>\n          </Adresse>\n        </Menge_Adresse>\n      </Patienten_Stammdaten>\n      <Menge_Meldung>\n        <Meldung Meldung_ID=\"ST0047290\" Melder_ID=\"ST\">\n          <Meldedatum>10.03.1900</Meldedatum>\n          <Meldebegruendung>I</Meldebegruendung>\n          <Meldeanlass>behandlungsbeginn</Meldeanlass>\n          <Tumorzuordnung Tumor_ID=\"1\">\n            <Primaertumor_ICD_Code>C61</Primaertumor_ICD_Code>\n            <Primaertumor_ICD_Version>10 2022 GM</Primaertumor_ICD_Version>\n            <Diagnosedatum>29.07.1900</Diagnosedatum>\n            <Seitenlokalisation>T</Seitenlokalisation>\n          </Tumorzuordnung>\n          <Menge_ST>\n            <ST ST_ID=\"361237\">\n              <ST_Intention>K</ST_Intention>\n              <ST_Stellung_OP>O</ST_Stellung_OP>\n              <Menge_Bestrahlung>\n                <Bestrahlung>\n                  <ST_Zielgebiet>5.4.+</ST_Zielgebiet>\n                  <ST_Seite_Zielgebiet>U</ST_Seite_Zielgebiet>\n                  <ST_Beginn_Datum>13.10.1900</ST_Beginn_Datum>\n                  <ST_Applikationsart>PRCN</ST_Applikationsart>\n                  <ST_Einzeldosis>\n                    <Dosis>1.8</Dosis>\n                    <Einheit>Gy</Einheit>\n                  </ST_Einzeldosis>\n                </Bestrahlung>\n                <Bestrahlung>\n                  <ST_Zielgebiet>5.4.-</ST_Zielgebiet>\n                  <ST_Seite_Zielgebiet>U</ST_Seite_Zielgebiet>\n                  <ST_Beginn_Datum>28.11.1900</ST_Beginn_Datum>\n                  <ST_Applikationsart>S</ST_Applikationsart>\n                </Bestrahlung>\n                <Bestrahlung>\n                  <ST_Zielgebiet>5.12.</ST_Zielgebiet>\n                  <ST_Seite_Zielgebiet>L</ST_Seite_Zielgebiet>\n                  <ST_Beginn_Datum>24.11.1900</ST_Beginn_Datum>\n                  <ST_Applikationsart>S</ST_Applikationsart>\n                </Bestrahlung>\n              </Menge_Bestrahlung>\n              <Anmerkung>Boost 1: Brachy-Boost in HDR-Technik ED 9,5 Gy GD 19 Gy Boost 2: SEQ ED 1,8\n                Gy GD 21,6 Gy</Anmerkung>\n            </ST>\n          </Menge_ST>\n        </Meldung>\n      </Menge_Meldung>\n    </Patient>\n  </Menge_Patient></ADT_GEKID>", "patient_id": "101777777777", "meldung_id": "UR0045297"},
    {"xml": "<?xml version='1.0' encoding='utf-8'?>\n<ADT_GEKID xmlns=\"http://www.gekid.de/namespace\" Schema_Version=\"2.2.1\"><Absender Absender_ID=\"UKKK_ONKOSTAR\" Software_ID=\"ONKOSTAR\" Installations_ID=\"2.11.1.4\">\n    <Absender_Bezeichnung>UKKK</Absender_Bezeichnung>\n    <Absender_Ansprechpartner />\n    <Absender_Anschrift>xxxxxxxxxxxxxxxxxxxx</Absender_Anschrift>\n  </Absender>\n  <Menge_Patient><Patient>\n      <Patienten_Stammdaten Patient_ID=\"101777777777\">\n        <KrankenversichertenNr>277777777</KrankenversichertenNr>\n        <KrankenkassenNr>9777777777</KrankenkassenNr>\n        <Patienten_Nachname>Riedelo</Patienten_Nachname>\n        <Patienten_Titel />\n        <Patienten_Namenszusatz />\n        <Patienten_Vornamen>M</Patienten_Vornamen>\n        <Patienten_Geburtsname />\n        <Patienten_Geschlecht>M</Patienten_Geschlecht>\n        <Patienten_Geburtsdatum>28.10.1900</Patienten_Geburtsdatum>\n        <Menge_Adresse>\n          <Adresse>\n            <Patienten_Strasse>xxxxxxxxx</Patienten_Strasse>\n            <Patienten_Hausnummer />\n            <Patienten_Land>DE</Patienten_Land>\n            <Patienten_PLZ>90</Patienten_PLZ>\n            <Patienten_Ort>xxxxxxx</Patienten_Ort>\n          </Adresse>\n        </Menge_Adresse>\n      </Patienten_Stammdaten>\n      <Menge_Meldung>\n        <Meldung Meldung_ID=\"ST0047290\" Melder_ID=\"ST\">\n          <Meldedatum>10.03.1900</Meldedatum>\n          <Meldebegruendung>I</Meldebegruendung>\n          <Meldeanlass>behandlungsbeginn</Meldeanlass>\n          <Tumorzuordnung Tumor_ID=\"1\">\n            <Primaertumor_ICD_Code>C61</Primaertumor_ICD_Code>\n            <Primaertumor_ICD_Version>10 2022 GM</Primaertumor_ICD_Version>\n            <Diagnosedatum>29.07.1900</Diagnosedatum>\n            <Seitenlokalisation>T</Seitenlokalisation>\n          </Tumorzuordnung>\n          <Menge_ST>\n            <ST ST_ID=\"361237\">\n              <ST_Intention>K</ST_Intention>\n              <ST_Stellung_OP>O</ST_Stellung_OP>\n              <Menge_Bestrahlung>\n                <Bestrahlung>\n                  <ST_Zielgebiet>5.4.+</ST_Zielgebiet>\n                  <ST_Seite_Zielgebiet>U</ST_Seite_Zielgebiet>\n                  <ST_Beginn_Datum>13.10.1900</ST_Beginn_Datum>\n                  <ST_Applikationsart>PRCN</ST_Applikationsart>\n                  <ST_Einzeldosis>\n                    <Dosis>1.8</Dosis>\n                    <Einheit>Gy</Einheit>\n                  </ST_Einzeldosis>\n                </Bestrahlung>\n                <Bestrahlung>\n                  <ST_Zielgebiet>5.4.-</ST_Zielgebiet>\n                  <ST_Seite_Zielgebiet>U</ST_Seite_Zielgebiet>\n                  <ST_Beginn_Datum>28.11.1900</ST_Beginn_Datum>\n                  <ST_Applikationsart>S</ST_Applikationsart>\n                </Bestrahlung>\n                <Bestrahlung>\n                  <ST_Zielgebiet>5.12.</ST_Zielgebiet>\n                  <ST_Seite_Zielgebiet>L</ST_Seite_Zielgebiet>\n                  <ST_Beginn_Datum>24.11.1900</ST_Beginn_Datum>\n                  <ST_Applikationsart>S</ST_Applikationsart>\n                </Bestrahlung>\n              </Menge_Bestrahlung>\n              <Anmerkung>Boost 1: Brachy-Boost in HDR-Technik ED 9,5 Gy GD 19 Gy Boost 2: SEQ ED 1,8\n                Gy GD 21,6 Gy</Anmerkung>\n            </ST>\n          </Menge_ST>\n        </Meldung>\n      </Menge_Meldung>\n    </Patient>\n  </Menge_Patient></ADT_GEKID>", "patient_id": "101777777777", "meldung_id": "ST0047290"},
  ])
# ---
